### PR TITLE
Implement exchange() function and a test.

### PR DESCRIPTION
**Pull request by vivekvpandya**

### CODE REVIEW

**Comment by andor0** [comment]
Could you run `cargo check -p picasso-runtime` to decrease changes of `Cargo.lock` file?

File: frame/curve-amm/src/lib.rs [hunk 1]
```diff
@@ -484,6 +504,102 @@ pub mod pallet {
...
+							.ok()
+							.ok_or(Error::<T>::Math)?,
+					);
+					let ann = Self::get_ann(amp_f, n_coins).ok_or(Error::<T>::Math)?;
+					let y = Self::get_y(i, j, x, &xp, ann).ok_or(Error::<T>::Math)?;
+
+					// -1 just in case there were some rounding errors
+					// dy = xp[j] - y - 1
+					let dy_f =
+						(|| xp[j].checked_sub(&y)?.checked_sub(&prec))().ok_or(Error::<T>::Math)?;
```
Comment (andor0) [inline frame/curve-amm/src/lib.rs:558]: Why does lambda need here?
imho, it is faster
```
let dy_f = xp[j].checked_sub(&y)
  .ok_or(Error::<T>::Math)?
  .checked_sub(&prec)
  .ok_or(Error::<T>::Math)?;
```

**Review by andor0 — CHANGES_REQUESTED** [review]

File: frame/curve-amm/src/lib.rs [hunk 2]
```diff
@@ -217,6 +217,26 @@ pub mod pallet {
...
+		/// - index of sent token `PoolTokenIndex`
+		/// - amount of sent token `T::Balance`
+		/// - index of received token `PoolTokenIndex`
+		/// - amount of received token `T::Balance`
+		///
+		/// \[who, pool_id, sent_token_index, sent_amount, received_token_index, received_amount\]
+		TokenExchange {
+			who: T::AccountId,
+			pool_id: PoolId,
+			sent_token_index: PoolTokenIndex,
```
[the remark this answers is not in the document]
Comment (vivekvpandya) [reply frame/curve-amm/src/lib.rs:235]: all other functions already have names related to math behind it. So I suggest to keep it this way.

[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply frame/curve-amm/src/lib.rs:235]: so extrinsic inputs has math names, but event deposits are adhock names?

File: frame/composable-traits/src/dex.rs [hunk 3]
```diff
@@ -122,6 +122,20 @@ pub trait CurveAmm {
...
+
+	/// Perform an exchange between two coins.
+	/// `i` - index value of the coin to send,
+	/// `j` - index value of the coin to recieve,
+	/// `dx` - amount of `i` being exchanged,
+	/// `min_dy` - minimum amount of `j` to receive.
+	fn exchange(
+		who: &Self::AccountId,
+		pool_id: PoolId,
+		i: PoolTokenIndex,
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply frame/composable-traits/src/dex.rs:134]: i see. so i think about
1. multi curve amm (multi dex) consis
2. consistent naming of input and naming of outputs

File: frame/curve-amm/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (vivekvpandya) [reply frame/curve-amm/src/lib.rs:235]: Events are externally observable so I prefer to put more descriptive names.

[the remark this answers is not in the document]
Comment (vivekvpandya) [reply frame/curve-amm/src/lib.rs:235]: note there is not extrinsic functions defined yet, We may choose to use more descriptive names in extrinsic definition.

File: frame/composable-traits/src/dex.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (vivekvpandya) [reply frame/composable-traits/src/dex.rs:134]: okay so if we really don't like these names then I think I need to do separate PR.  which will rename functions like get_y(), get_dy() etc too

**Review by dzmitry-lahoda — DISMISSED** [review]

File: frame/curve-amm/src/lib.rs [hunk 4]
```diff
@@ -484,6 +504,105 @@ pub mod pallet {
...
+			dx: Self::Balance,
+			min_dy: Self::Balance,
+		) -> Result<(), DispatchError> {
+			let prec = T::Precision::get();
+			let zero_b = Self::Balance::zero();
+			ensure!(dx >= zero_b, Error::<T>::WrongAssetAmount);
+
+			let (provider, pool_id, dy) =
+				Pools::<T>::try_mutate(pool_id, |pool| -> Result<_, DispatchError> {
+					let pool = pool.as_mut().ok_or(Error::<T>::PoolNotFound)?;
```
Comment (hussein-aitlahcen) [inline frame/curve-amm/src/lib.rs:521]: Why do we have a `as_mut` here? The `pool` should be a `&mut ...` considering we are using `try_mutate`, unless I missed something?

**Review by hussein-aitlahcen — DISMISSED** [review]
I think we should remove assets, balances and total_balances from PoolInfo and use proper storage maps to represent them. PoolInfo would then just be the pool configuration, and all it's state would be stored accross multiple (or one?) storagemap indexed by the (pool_id, asset_id)? I'm not sure using Vec is idiomatic here, they probably did an almost 1:1 with their solidity contracts

File: frame/curve-amm/src/lib.rs (hunk 4 above)
Comment (vivekvpandya) [reply frame/curve-amm/src/lib.rs:521]: try_mutate() returns &mut Option<T> so then with as_mut it is converted to Option<&mut T>

**Review by dzmitry-lahoda — APPROVED** [review]

**Review by hussein-aitlahcen — APPROVED** [review]

**Pull request merged**